k2/csrc/fsa_algo.cu

```diff
@@ -747,4 +747,100 @@ Fsa Union(FsaVec &fsas, Array1<int32_t> *arc_map /*= nullptr*/) {
   return ans;
 }
 
+Fsa Closure(Fsa &fsa, Array1<int32_t> *arc_map /* = nullptr*/) {
+  K2_CHECK_EQ(fsa.NumAxes(), 2) << "We support only a single FSA.";
+  ContextPtr &c = fsa.Context();
+
+  int32_t num_states = fsa.Dim0();
+  if (num_states < 2) {
+    K2_CHECK_EQ(num_states, 0)
+        << "An empty fsa should contain no states at all";
+    if (arc_map != nullptr) *arc_map = Array1<int32_t>(c, 0);
+    return fsa;  // return itself if the input fsa is empty
+  }
+
+  const int32_t *fsa_row_splits_data = fsa.RowSplits(1).Data();
+  const int32_t *fsa_row_ids_data = fsa.RowIds(1).Data();
+  const Arc *fsa_arcs_data = fsa.values.Data();
+  int32_t fsa_final_state = num_states - 1;
+
+  int32_t num_out_states = num_states;
+
+  // An arc from the start state to the final state with label == 0 is added.
+  int32_t num_out_arcs = fsa.values.Dim() + 1;
+
+  Array1<int32_t> out_row_ids(c, num_out_arcs);
+  int32_t *out_row_ids_data = out_row_ids.Data();
+
+  Array1<Arc> out_arcs(c, num_out_arcs);
+  Arc *out_arcs_data = out_arcs.Data();
+
+  Array1<int32_t> tmp_arc_map(c, num_out_arcs, -1);
```

Comment:
it may not need to init with -1 if you set arc map data for the added arc from start_state to final_state

Reply:
yes, you are right.